### ISSUE TITLE
IAMPartialPolicy CRD validation fails on valid resource

### ISSUE THREAD

**Issue by jgogstad** [GoogleCloudPlatform/k8s-config-connector#776]
### Checklist

- [X] I did not find a related open issue.
- [X] I did not find a solution in the troubleshooting guide: (https://cloud.google.com/config-connector/docs/troubleshooting)
- [X] If this issue is time-sensitive, I have submitted a corresponding issue with [GCP support](https://cloud.google.com/support-hub).

### Bug Description

The CRD validation for IAMPartialPolicy seems broken:

```
❯ kubectl apply -f - <<< '---
apiVersion: iam.cnrm.cloud.google.com/v1beta1
kind: IAMPartialPolicy
metadata:
  name: pubsub-subscribers
  namespace: my-ns
spec:
  bindings:
  - members:
    - member: "serviceAccount:foo@bar.iam.gserviceaccount.com"
    role: roles/pubsub.subscriber
  resourceRef:
    kind: Project
...'
The IAMPartialPolicy "pubsub-subscribers-2" is invalid:
* <nil>: Invalid value: "": "spec.resourceRef" must validate one and only one schema (oneOf). Found none valid
* spec.resourceRef.name: Required value
❯
```

It should pass this CRD validation, but it doesn't: https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/v1.89.0/crds/iam_v1beta1_iampartialpolicy.yaml#L168

If I add `resourceRef.name`, then it works.

### Additional Diagnostic Information

<details>
<summary>kubectl get namespace my-ns -o yaml</summary>

```yaml
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    cnrm.cloud.google.com/project-id: my-project
    kubectl.kubernetes.io/last-applied-configuration: |
      # snip
  creationTimestamp: "2023-01-11T21:46:58Z"
  labels:
    kubernetes.io/metadata.name: my-project
  name: bar
  resourceVersion: "104632170"
  uid: f72a099f-2e3c-425e-be82-965b386d0893
spec:
  finalizers:
  - kubernetes
status:
  phase: Active
```
</details>

### Kubernetes Cluster Version

❯ k version --short Flag --short has been deprecated, and will be removed in the future. The --short output will become the default. Client Version: v1.26.1 Kustomize Version: v4.5.7 Server Version: v1.24.9-gke.2000

### Config Connector Version

1.89.0

### Config Connector Mode

namespaced mode (default)

### Steps to reproduce the issue

Run same `kubectl apply` as in description

**Comment by diviner524** [comment]
@jgogstad The link you shared (with the rule below) is from some deprecated feature (never documented anywhere) that needs to be cleaned up. The initial concept was for users to specify only the kind as `Project`, allowing Config Connector to automatically derive the project ID from the `project-id` annotation in the corresponding Kubernetes namespace. However, this approach has proven to be confusing and is not an ideal solution. 

We no longer support this feature and are planning to remove these legacy implementations.

```
                - not:
                    anyOf:
                    - required:
                      - name
                    - required:
                      - namespace
                    - required:
                      - apiVersion
                    - required:
                      - external
```

In your case, you could either use `name` (and `namespace`) to refer to a GCP project K8s resource, or just use `external` to specify a project id explicitly.

**Comment by jgogstad** [comment]
@diviner524 thank you for explaining, that's helpful

> The link you shared (with the rule below) is from some deprecated feature (never documented anywhere) that needs to be cleaned up.
>
> The initial concept was for users to specify only the kind as Project, allowing Config Connector to automatically derive the project ID from the project-id annotation in the corresponding Kubernetes namespace. However, this approach has proven to be confusing and is not an ideal solution.
>
> We no longer support this feature and are planning to remove these legacy implementations.

I think your reply contradicts the current documented public API. I understand that the CRD I linked may not be in use, but the feature itself is definitely documented to behave in this way:

https://cloud.google.com/config-connector/docs/how-to/organizing-resources/project-scoped-resources :

> In order for Config Connector to determine where to create your resources, it checks for an annotation on your resource. If there is no annotation applied to the resource, Config Connector checks to see if the namespace has a targeted-scope annotation.
> …
> To set a default project ID for newly created resources, annotate your namespace.

https://cloud.google.com/config-connector/docs/concepts/namespaces-and-projects :

> Config Connector can use the annotations on the resource's Namespace to determine what project, folder, or organization to create the resources in.

https://cloud.google.com/config-connector/docs/how-to/organizing-resources/overview

> For resources that do not supply this annotation in their configuration, Config Connector will check the namespace annotations. You can default the annotation for all resources in a particular namespace by annotating the namespace. The namespace annotation will only be applied to resources that both exist in the namespace and do not have the scope-defining annotation directly on the resource.

Regarding this part of your comment:

> We no longer support this feature and are planning to remove these legacy implementations.

Based on the linked documentation above, this is clearly an API breaking change. From a user perspective, I have resources applied to my GKE that I can no longer re-apply without getting that error in original message. How can I stay on top of API breaking changes in config connector?

**Comment by diviner524** [comment]
@jgogstad All the links you shared are still valid and we still support these features (including namespace level annotation for `project-id`). However do note we are discussing a very different CRD, in this case `IAMPartialPolicy`.

IAMPartialPolicy is attached to a GCP resource, but not a project, so we do not need to specify a project id. This is also reflected in the resource reference page [1] as IAMPartialPolicy doesn't support `cnrm.cloud.google.com/project-id` there. It just happens in your case the "resource" is a GCP project. It can be a project, folder, organization, or ContainerCluster, any resources that supports IAM.

As long as a correct resourceRef is provided, the IAM policy will be configured through the `setIamPolicy` method of that resource, the actual project id (or for folder/organization there is simply no project involved) of that resource is not really needed. 

In other words, the description below does not apply to IAMPolicy, as Config Connector uses the resourceRef to determine where to configure the IAMPolicy, instead of annotation. We will update our documentation to better explain the difference.

> In order for Config Connector to determine where to create your resources, it checks for an annotation on your resource.


[1] https://cloud.google.com/config-connector/docs/reference/resource-docs/iam/iampartialpolicy#annotations

**Comment by jgogstad** [comment]
@diviner524 

> All the links you shared are still valid and we still support these features (including namespace level annotation for project-id)

That's good to hear, something wasn't quite right.

> In other words, the description below does not apply to IAMPolicy, as Config Connector uses the resourceRef to determine where to configure the IAMPolicy, instead of annotation. We will update our documentation to better explain the difference.
> 
>   >  In order for Config Connector to determine where to create your resources, it checks for an annotation on your resource.

Right, this is the confusing part. It's the sentence after the one you quoted that is the main source of confusion:

> If there is no annotation applied to the resource, Config Connector checks to see if the namespace has a targeted-scope annotation.

If I understand what you're saying correctly, this sentence does simply not apply to `IAMPartialPolicy`. Could you confirm that?

**Comment by jgogstad** [comment]
No matter how you look at this though, you have a regression in your service. Something that used to work no longer works; downstream broke because you made a change. I literally have `IAMPartialPolicy` objects deployed that I can't redeploy without hitting schema validation errors.

**Comment by diviner524** [comment]
@jgogstad

> If I understand what you're saying correctly, this sentence does simply not apply to IAMPartialPolicy. Could you confirm that?

Yes, and we will update our documentation to make this clear.

And I agree it is a breaking change if some users have discovered and been using IAMPartialPolicy (for Project resources) this way. We will try to avoid this in future.

**Comment by jgogstad** [comment]
Thank you for your feedback. You've made it clear how to resolve this issue, so closing this ticket. Feel free to reopen if you want to use it for tracking the documentation updates.
